File: water-guardian-simulation/firmware/esp32_prototype/src/rule_engine.cpp
#include "rule_engine.h"
#include <string.h>

namespace wg {

RuleEngine::RuleEngine(const Rules& r) : R_(r) { memset(ev_, 0, sizeof(ev_)); }

const char* RuleEngine::stateName(State s) {
  static const char* n[] = {"NORMAL", "EXPECTED_FLOW", "PERSISTENT_FLOW", "UNEXPECTED_FLOW", "REVIEW", "SENSOR_UNKNOWN", "RESOLVED"};
  return n[(int)s];
}
const char* RuleEngine::modeName(Mode m) {
  static const char* n[] = {"QUIET", "PRAYER", "CLEANING", "SPECIAL_EVENT"};
  return n[(int)m];
}

void RuleEngine::emit(EventType t, float f0, float f1, uint8_t u0) {
  Event e{nextId_++, t_, t, f0, f1, u0, false};
  uint16_t idx = (evHead_ + evLen_) % EV;
  if (evLen_ == EV) { evHead_ = (evHead_ + 1) % EV; idx = (evHead_ + EV - 1) % EV; } else evLen_++;
  ev_[idx] = e;
}

uint16_t RuleEngine::pendingEvents(Event* out, uint16_t max) const {
  uint16_t n = 0;
  for (uint16_t i = 0; i < evLen_ && n < max; i++) { const Event& e = ev_[(evHead_ + i) % EV]; if (!e.synced) out[n++] = e; }
  return n;
}
void RuleEngine::markSynced(uint32_t uptoId) {
  for (uint16_t i = 0; i < evLen_; i++) { Event& e = ev_[(evHead_ + i) % EV]; if (e.id <= uptoId) e.synced = true; }
}
uint16_t RuleEngine::unsyncedCount() const { uint16_t n = 0; for (uint16_t i = 0; i < evLen_; i++) if (!ev_[(evHead_ + i) % EV].synced) n++; return n; }

void RuleEngine::save(Persisted& p) const {
  p.magic = 0x57475631; p.incident = inc_; p.nextId = nextId_; p.aboveSinceS = aboveSince_;
  p.nUnsynced = pendingEvents(p.unsynced, 32);
}
void RuleEngine::restore(const Persisted& p) {
  if (p.magic != 0x57475631) return;
  inc_ = p.incident; nextId_ = p.nextId;
  evHead_ = 0; evLen_ = 0;
  for (uint16_t i = 0; i < p.nUnsynced && i < 32; i++) { ev_[evLen_++] = p.unsynced[i]; }
  // After a reboot the monotonic clock restarts: we cannot know how long the flow was above
  // threshold before the reset, so we keep the incident open (conservative) and restart timers.
  aboveSince_ = inc_.active ? 0 : -1;
  if (inc_.active) state_ = inc_.priority == Priority::HIGH ? State::UNEXPECTED_FLOW : State::REVIEW;
  emit(EventType::BOOT, inc_.active ? 1.f : 0.f, (float)p.nUnsynced);
}

bool RuleEngine::action(Action a) {
  if (!inc_.active) return false;
  inc_.acked = true; inc_.ackType = a;
  emit(EventType::ACTION, (float)inc_.id, 0, (uint8_t)a);
  if (a == Action::EXPECTED_ACTIVITY) { inc_.priority = Priority::EXPECTED; state_ = State::EXPECTED_FLOW; }
  return true;
}

void RuleEngine::openIncident(Priority p, float q, float aboveMin) {
  inc_ = Incident{};
  inc_.active = true; inc_.id = nextId_; inc_.openedAtS = t_; inc_.mode = effMode_; inc_.priority = p;
  inc_.peakLpm = q; inc_.lastFlowLpm = q; inc_.litres = q * aboveMin; inc_.durationMin = aboveMin; inc_.quality = lastQuality_;
  state_ = p == Priority::HIGH ? State::UNEXPECTED_FLOW : State::REVIEW;
  emit(p == Priority::HIGH ? EventType::UNEXPECTED_FLOW : EventType::REVIEW, q, aboveMin, (uint8_t)effMode_);
}

State RuleEngine::step(const Sample& s, const Context& c) {
  const uint32_t t = s.tSec;
  const float dt = t > t_ ? (float)(t - t_) : 0.f;
  t_ = t;

  // ---- effective facility mode ----
  if (c.mode == Mode::CLEANING) {
    if (cleaningStarted_ < 0) { cleaningStarted_ = (int32_t)t; cleaningExpiredLogged_ = false; emit(EventType::CLEANING_START); }
    bool expired = (t - cleaningStarted_) / 60.f >= R_.cleaningOverrideMin;
    effMode_ = expired ? c.underlyingMode : Mode::CLEANING;
    if (expired && !cleaningExpiredLogged_) { cleaningExpiredLogged_ = true; emit(EventType::CLEANING_OVERRIDE_EXPIRED); }
  } else { cleaningStarted_ = -1; effMode_ = c.mode; }
  if (!c.clockValid) effMode_ = R_.clockInvalidTreatAsQuiet ? Mode::QUIET : Mode::PRAYER;
  const bool quiet = effMode_ == Mode::QUIET, cleaning = effMode_ == Mode::CLEANING;

  // ---- validity (Principle 1) ----
  const bool valid = s.commOk && s.valid && s.quality >= R_.qualityThreshold;
  if (!valid) {
    if (invalidSince_ < 0) invalidSince_ = (int32_t)t;
    if (t - invalidSince_ >= R_.sensorUnknownGraceS && state_ != State::SENSOR_UNKNOWN) {
      state_ = State::SENSOR_UNKNOWN; emit(EventType::SENSOR_UNKNOWN, 0, 0, inc_.active ? 1 : 0);
    }
    return state_;  // timers and incident untouched: we do not know.
  }
  if (invalidSince_ >= 0) {
    invalidSince_ = -1;
    if (state_ == State::SENSOR_UNKNOWN) { emit(EventType::SENSOR_RECOVERED); state_ = inc_.active ? (inc_.priority == Priority::HIGH ? State::UNEXPECTED_FLOW : State::REVIEW) : State::NORMAL; }
  }

  // ---- moving average ----
  uint16_t idx = (winHead_ + winLen_) % WIN;
  if (winLen_ == WIN) { winHead_ = (winHead_ + 1) % WIN; idx = (winHead_ + WIN - 1) % WIN; } else winLen_++;
  winQ_[idx] = s.flowLpm; winT_[idx] = t;
  while (winLen_ > 1 && t - winT_[winHead_] > R_.averagingWindowS) { winHead_ = (winHead_ + 1) % WIN; winLen_--; }
  float sum = 0; for (uint16_t i = 0; i < winLen_; i++) sum += winQ_[(winHead_ + i) % WIN];
  const float q = sum / winLen_;
  lastFlow_ = q; lastQuality_ = s.quality;

  const bool above = q > R_.flowThresholdLpm;
  const bool belowResolve = q < R_.flowThresholdLpm * (1.f - R_.hysteresisPct / 100.f);

  if (inc_.active && above) { inc_.litres += q / 60.f * dt; if (q > inc_.peakLpm) inc_.peakLpm = q; inc_.lastFlowLpm = q; inc_.durationMin = (t - (aboveSince_ >= 0 ? (uint32_t)aboveSince_ : inc_.openedAtS)) / 60.f; }

  if (above) { if (aboveSince_ < 0) aboveSince_ = (int32_t)t; belowSince_ = -1; }
  else if (belowResolve) { if (belowSince_ < 0) belowSince_ = (int32_t)t; }
  if (!above && belowSince_ >= 0 && !inc_.active && t - belowSince_ >= R_.gapToleranceS) aboveSince_ = -1;
  const float aboveMin = this->aboveMin();

  // ---- resolution ----
  if (inc_.active && belowSince_ >= 0 && (t - belowSince_) / 60.f >= R_.resolveMin) {
    emit(EventType::INCIDENT_RESOLVED, inc_.litres, inc_.durationMin, (uint8_t)inc_.ackType);
    inc_.active = false; state_ = State::RESOLVED; resolvedAt_ = (int32_t)t; aboveSince_ = -1;
    return state_;
  }
  if (state_ == State::RESOLVED) {
    if (above) state_ = State::NORMAL;
    else if ((t - resolvedAt_) / 60.f >= R_.resolvedDisplayMin) state_ = State::NORMAL;
    else return state_;
  }

  // ---- decision table ----
  if (inc_.active) { state_ = inc_.priority == Priority::HIGH ? State::UNEXPECTED_FLOW : inc_.priority == Priority::EXPECTED ? State::EXPECTED_FLOW : State::REVIEW; return state_; }
  if (!above) { state_ = State::NORMAL; return state_; }
  if (cleaning) { state_ = State::EXPECTED_FLOW; return state_; }
  if (quiet) { if (aboveMin >= R_.persistenceMinQuiet) openIncident(Priority::HIGH, q, aboveMin); else state_ = State::PERSISTENT_FLOW; return state_; }
  if (aboveMin >= R_.persistenceMinOccupied) openIncident(Priority::LOW, q, aboveMin); else state_ = State::EXPECTED_FLOW;
  return state_;
}

}  // namespace wg

File: water-guardian-simulation/simulator/engine/ruleEngine.js
/* Water Guardian — persistent-unexpected-flow RULE ENGINE (state machine).
 *
 * Design principles (mirrors firmware/esp32_prototype/src/rule_engine.cpp):
 *  1. INVALID SENSOR DATA IS NEVER ZERO FLOW  → SENSOR_UNKNOWN.
 *  2. Detection is purely local (no network dependency). Connectivity is an
 *     ORTHOGONAL "link" state; events are buffered until they can sync.
 *  3. Every threshold is a configurable parameter, not a fact.
 *  4. The engine never claims "LEAK"; it claims "persistent unexpected flow".
 *
 * States: NORMAL · EXPECTED_FLOW · PERSISTENT_FLOW · UNEXPECTED_FLOW · REVIEW ·
 *         SENSOR_UNKNOWN · RESOLVED   (+ link: ONLINE / OFFLINE overlay)
 *
 * Works in browser (window.WG) and Node (module.exports).
 */
(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) module.exports = factory();
  else root.WG = Object.assign(root.WG || {}, factory());
})(typeof self !== 'undefined' ? self : this, function () {
  const STATES = ['NORMAL', 'EXPECTED_FLOW', 'PERSISTENT_FLOW', 'UNEXPECTED_FLOW', 'REVIEW', 'SENSOR_UNKNOWN', 'RESOLVED'];
  const MODES = ['QUIET', 'PRAYER', 'CLEANING', 'SPECIAL_EVENT'];

  const DEFAULT_RULES = {
    flowThresholdLpm: 0.5,         // valid flow above this counts as "flow present"   (ASSUMED default)
    hysteresisPct: 20,             // resolve when flow < threshold × (1 − h)           (ASSUMED)
    persistenceMinQuiet: 10,       // QUIET: minutes above threshold → UNEXPECTED_FLOW (ASSUMED)
    persistenceMinOccupied: 45,    // PRAYER/SPECIAL_EVENT: minutes → REVIEW (low priority)
    cleaningOverrideMin: 60,       // CLEANING override auto-expires after this
    resolveMin: 3,                 // minutes below (threshold − hysteresis) to resolve
    gapToleranceS: 30,             // a dip below threshold shorter than this does not reset persistence
    averagingWindowS: 30,          // moving average of VALID samples before threshold comparison
    qualityThreshold: 60,          // meter Q below this → sample INVALID
    sensorUnknownGraceS: 15,       // consecutive invalid seconds before SENSOR_UNKNOWN
    resolvedDisplayMin: 5,         // how long RESOLVED is shown before NORMAL
    clockInvalidPolicy: 'OCCUPIED',// if RTC/NTP invalid: treat as OCCUPIED (conservative) or QUIET
  };

  function createEngine(rules, opts) {
    const R = Object.assign({}, DEFAULT_RULES, rules || {});
    const O = Object.assign({ now: () => 0 }, opts || {});
    let s = freshState();

    function freshState() {
      return {
        state: 'NORMAL',
        link: 'ONLINE',
        mode: 'QUIET',
        effectiveMode: 'QUIET',
        clockValid: true,
        aboveSinceS: null,        // sim-time when valid flow first exceeded threshold (continuous)
        belowSinceS: null,        // sim-time when valid flow first fell below resolve level
        invalidSinceS: null,
        cleaningStartedS: null,
        resolvedAtS: null,
        lastValidFlowLpm: null,
        lastRawFlowLpm: null,
        lastQuality: null,
        window: [],
        lastSample: null,
        incident: null,           // {id, openedAtS, mode, peakLpm, litres, maxDurationMin, priority, acked, ackType}
        events: [],               // {id, tSec, type, payload, synced}
        nextId: 1,
        tSec: 0,
      };
    }

    /** Persist/restore — what the ESP32 keeps in NVS across a reboot. */
    function snapshot() {
      return JSON.stringify({
        incident: s.incident,
        events: s.events.filter((e) => !e.synced),
        nextId: s.nextId,
        aboveSinceS: s.aboveSinceS,
        cleaningStartedS: s.cleaningStartedS,
      });
    }
    function restore(json) {
      const d = JSON.parse(json);
      s = freshState();
      s.incident = d.incident || null;
      s.events = d.events || [];
      s.nextId = d.nextId || 1;
      s.aboveSinceS = d.aboveSinceS ?? null;
      s.cleaningStartedS = d.cleaningStartedS ?? null;
      if (s.incident) s.state = s.incident.priority === 'HIGH' ? 'UNEXPECTED_FLOW' : 'REVIEW';
      emit('BOOT', { restoredIncident: !!s.incident, unsyncedEvents: s.events.length });
    }
    function reboot() { const snap = snapshot(); restore(snap); }

    function emit(type, payload) {
      const ev = { id: s.nextId++, tSec: s.tSec, type, payload: payload || {}, synced: false };
      s.events.push(ev);
      if (s.events.length > 500) s.events.shift();
      return ev;
    }

    /** Try to sync unsynced events; returns number synced. */
    function sync(canReachBackend) {
      if (!canReachBackend) return 0;
      let n = 0;
      for (const e of s.events) if (!e.synced) { e.synced = true; n++; }
      return n;
    }

    /**
     * Advance the engine by one sample.
     * @param {object} sample  {tSec, commOk, valid, flowLpm, quality, workingStep, invalidReason}
     * @param {object} ctx     {mode, clockValid, wifi, backend}
     */
    function step(sample, ctx) {
      const t = sample.tSec;
      const dt = Math.max(0, t - s.tSec);
      s.tSec = t;
      s.lastSample = sample;
      s.mode = ctx.mode;
      s.clockValid = ctx.clockValid !== false;

      // ---- link overlay (never gates detection) ----
      const newLink = ctx.wifi && ctx.backend ? 'ONLINE' : 'OFFLINE';
      if (newLink !== s.link) { s.link = newLink; emit('LINK', { link: newLink }); }
      if (s.link === 'ONLINE') sync(true);

      // ---- effective facility mode ----
      if (ctx.mode === 'CLEANING') {
        if (s.cleaningStartedS === null) { s.cleaningStartedS = t; emit('CLEANING_START', {}); }
        const expired = (t - s.cleaningStartedS) / 60 >= R.cleaningOverrideMin;
        s.effectiveMode = expired ? (ctx.underlyingMode || 'QUIET') : 'CLEANING';
        if (expired && s._cleaningExpiredLogged !== s.cleaningStartedS) { s._cleaningExpiredLogged = s.cleaningStartedS; emit('CLEANING_OVERRIDE_EXPIRED', {}); }
      } else {
        s.cleaningStartedS = null;
        s.effectiveMode = ctx.mode;
      }
      if (!s.clockValid) s.effectiveMode = R.clockInvalidPolicy === 'QUIET' ? 'QUIET' : 'PRAYER';
      const quiet = s.effectiveMode === 'QUIET';
      const cleaning = s.effectiveMode === 'CLEANING';

      // ---- sample validity (Principle 1) ----
      const valid = !!sample.commOk && !!sample.valid && sample.flowLpm !== null && (sample.quality ?? 0) >= R.qualityThreshold;
      if (!valid) {
        if (s.invalidSinceS === null) s.invalidSinceS = t;
        const unknownFor = t - s.invalidSinceS;
        if (unknownFor >= R.sensorUnknownGraceS && s.state !== 'SENSOR_UNKNOWN') {
          s.state = 'SENSOR_UNKNOWN';
          emit('SENSOR_UNKNOWN', { reason: sample.invalidReason || (sample.commOk ? 'INVALID' : 'NO_COMM'), incidentOpen: !!s.incident });
        }
        // Do NOT touch aboveSince/belowSince or incident: we simply do not know.
        return view();
      }
      if (s.invalidSinceS !== null) {
        s.invalidSinceS = null;
        if (s.state === 'SENSOR_UNKNOWN') {
          emit('SENSOR_RECOVERED', {});
          s.state = s.incident ? (s.incident.priority === 'HIGH' ? 'UNEXPECTED_FLOW' : 'REVIEW') : 'NORMAL';
        }
      }
      // Moving average of valid samples (firmware does the same with a ring buffer).
      s.window.push({ t, q: sample.flowLpm });
      while (s.window.length && t - s.window[0].t > R.averagingWindowS) s.window.shift();
      const q = s.window.reduce((a, w) => a + w.q, 0) / s.window.length;
      s.lastValidFlowLpm = q;
      s.lastRawFlowLpm = sample.flowLpm;
      s.lastQuality = sample.quality;

      const above = q > R.flowThresholdLpm;
      const belowResolve = q < R.flowThresholdLpm * (1 - R.hysteresisPct / 100);

      // ---- incident accounting ----
      if (s.incident && above) {
        s.incident.litres += (q / 60) * dt;
        s.incident.peakLpm = Math.max(s.incident.peakLpm, q);
        s.incident.lastFlowLpm = q;
        s.incident.durationMin = (t - (s.aboveSinceS ?? s.incident.openedAtS)) / 60;
      }

      // ---- above/below timers ----
      if (above) {
        if (s.aboveSinceS === null) s.aboveSinceS = t;
        s.belowSinceS = null;
      } else if (belowResolve) {
        if (s.belowSinceS === null) s.belowSinceS = t;
      }
      // In the hysteresis band we hold both timers (no change).
      // Persistence (no incident yet) resets after a dip longer than gapToleranceS;
      // an OPEN incident only resolves after resolveMin (handled below).
      if (!above && s.belowSinceS !== null && !s.incident && t - s.belowSinceS >= R.gapToleranceS) {
        s.aboveSinceS = null;
      }

      const aboveMin = s.aboveSinceS === null ? 0 : (t - s.aboveSinceS) / 60;

      // ---- resolution ----
      if (s.incident && s.belowSinceS !== null && (t - s.belowSinceS) / 60 >= R.resolveMin) {
        const inc = s.incident;
        emit('INCIDENT_RESOLVED', { id: inc.id, litres: round(inc.litres, 1), durationMin: round(inc.durationMin, 1), ackType: inc.ackType });
        s.incident = null;
        s.state = 'RESOLVED';
        s.resolvedAtS = t;
        s.aboveSinceS = null;
        return view();
      }
      if (s.state === 'RESOLVED') {
        if (above) s.state = 'NORMAL'; // fall through to re-evaluate below
        else if ((t - s.resolvedAtS) / 60 >= R.resolvedDisplayMin) s.state = 'NORMAL';
        else return view();
      }

      // ---- main decision table ----
      if (s.incident) {
        // Incident stays open until resolved (flow below) — mode changes do not close it.
        s.state = s.incident.priority === 'HIGH' ? 'UNEXPECTED_FLOW' : 'REVIEW';
        return view();
      }
      if (!above) { s.state = 'NORMAL'; return view(); }

      if (cleaning) { s.state = 'EXPECTED_FLOW'; return view(); }

      if (quiet) {
        if (aboveMin >= R.persistenceMinQuiet) openIncident('HIGH', q, aboveMin);
        else s.state = 'PERSISTENT_FLOW';
        return view();
      }
      // PRAYER / SPECIAL_EVENT (or clock invalid, conservative)
      if (aboveMin >= R.persistenceMinOccupied) openIncident('LOW', q, aboveMin);
      else s.state = 'EXPECTED_FLOW';
      return view();
    }

    function openIncident(priority, q, aboveMin) {
      s.incident = {
        id: s.nextId, openedAtS: s.tSec, startedAtS: s.aboveSinceS, mode: s.effectiveMode, priority,
        peakLpm: q, lastFlowLpm: q, litres: (q / 60) * aboveMin * 60, durationMin: aboveMin,
        quality: s.lastQuality, acked: false, ackType: null,
        reason: priority === 'HIGH'
          ? `Flow remained above ${R.flowThresholdLpm} L/min for ${aboveMin.toFixed(0)} minutes during a QUIET period.`
          : `Flow remained above ${R.flowThresholdLpm} L/min for ${aboveMin.toFixed(0)} minutes during an OCCUPIED period (${s.effectiveMode}). Low-priority review.`,
      };
      s.state = priority === 'HIGH' ? 'UNEXPECTED_FLOW' : 'REVIEW';
      emit(priority === 'HIGH' ? 'UNEXPECTED_FLOW' : 'REVIEW', { id: s.incident.id, flowLpm: q, durationMin: round(aboveMin, 1), mode: s.effectiveMode, quality: s.lastQuality, reason: s.incident.reason });
    }

    /** Caretaker actions: ACKNOWLEDGE | EXPECTED_ACTIVITY | INCIDENT_CHECKED */
    function action(type) {
      if (!s.incident) return false;
      s.incident.acked = true;
      s.incident.ackType = type;
      emit('ACTION_' + type, { id: s.incident.id });
      if (type === 'EXPECTED_ACTIVITY') {
        // Human says this flow is legitimate: downgrade to EXPECTED_FLOW until it stops; do not re-alert.
        s.incident.priority = 'EXPECTED';
        s.state = 'EXPECTED_FLOW';
      }
      return true;
    }

    function view() {
      return {
        state: s.state, link: s.link, mode: s.mode, effectiveMode: s.effectiveMode, clockValid: s.clockValid,
        aboveMin: s.aboveSinceS === null ? 0 : (s.tSec - s.aboveSinceS) / 60,
        lastValidFlowLpm: s.lastValidFlowLpm, lastQuality: s.lastQuality,
        incident: s.incident ? Object.assign({}, s.incident) : null,
        unsynced: s.events.filter((e) => !e.synced).length,
        tSec: s.tSec,
      };
    }
    function events() { return s.events.slice(); }
    function rulesCopy() { return Object.assign({}, R); }
    function setRule(k, v) { R[k] = v; }

    return { step, action, view, events, snapshot, restore, reboot, rules: rulesCopy, setRule, STATES, MODES };
  }

  function round(x, n) { const f = Math.pow(10, n); return Math.round(x * f) / f; }
  return { STATES, MODES, DEFAULT_RULES, createEngine };
});
